### ISSUE TITLE
Warning about ambiguous provider preference

### ISSUE THREAD

**Issue by lunaryorn** [jssjr/git#64]
On Ubuntu 14.04 I get the following warning with `include_recipe 'git'`:

```
WARN: Ambiguous provider precedence: [Chef::Provider::GitClient::Package, Chef::Provider::GitClient::Source], please use Chef.set_provider_priority_array to provide determinism
```

How to fix this?

**Comment by signe** [comment]
I'm seeing this as well.  

```
Recipe: git::default
  * git_client[default] action install[2015-06-11T23:37:08+00:00] WARN: Ambiguous provider precedence: [Chef::Provider::GitClient::Package, Chef::Provider::GitClient::Source], please use Chef.set_provider_priority_array to provide determinism

    * apt_package[default :create git] action install (up to date)
     (up to date)
```

**Comment by lamont-granquist** [comment]
The problem is that both of those providers declare they `provide :git_client, os: 'linux'` and the choice of which to use is ambiguous.  12.0 through 12.3 is going to pick the Package provider because we do an alpha sort of the class names and pick the first in that code.  In 12.4 that necessarily will reverse and we will get last-class-declared wins which will be by alphabetical sort order of the library files so that the Source provider will win.

When 12.4 gets released this will likely result in people who were doing Package installs with this cookbook getting Source installs instead.

The solution is to:

```
Chef.set_provider_priority_array :git_client,  Chef::Provider::GitClient::Package, os: "linux"
```

This can also go directly into the provider library files:

e.g.

https://github.com/chef/chef/blob/master/lib/chef/provider/package.rb#L510-L523

**Comment by someara** [comment]
This should be fixed up in 4.3.2
